source/server/hot_restarting_child.h

```diff
@@ -14,6 +14,28 @@ class HotRestartingChild : HotRestartingBase {
   HotRestartingChild(int base_id, int restart_epoch, const std::string& socket_path,
                      mode_t socket_mode);
 
+  // A structure to record the set of registered UDP listeners keyed on their addresses,
+  // to support QUIC packet forwarding.
+  class UdpForwardingContext {
+  public:
+    using ForwardEntry = std::pair<const Network::Address::Instance*, Network::UdpListenerConfig*>;
+
+    // Returns the address and UdpListenerConfig associated with the given address.
+    // The addresses are not necessarily identical, as e.g. the listener might be listening on
+    // 0.0.0.0.
+    absl::optional<ForwardEntry>
+    getListenerForDestination(const Network::Address::Instance& address);
+
+    // Registers a UdpListenerConfig and address into the map, to be matched using
+    // getListenerForDestination for UDP packet forwarding.
+    void registerListener(const Network::Address::Instance& address,
+                          Network::UdpListenerConfig& listener_config);
+
+  private:
+    // Map keyed on address as a string, because Network::Address::Instance isn't hashable.
+    absl::flat_hash_map<std::string, ForwardEntry> listener_map_;
```

Comment:
Any reason using string as key type instead of `Network::Address::InstanceSharedPtr`?

Reply:
We don't have a SharedPtr, and when we need to do the lookup the address will be constructed from wire bytes so wouldn't be the same instance even if we did.

Reply:
ACK